dali/pipeline/operators/detection/box_encoder.cu

```diff
@@ -0,0 +1,264 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/detection/box_encoder.cuh"
+#include <cuda.h>
+#include <vector>
+#include <utility>
+
+namespace dali {
+__host__ __device__ float4 ToCenterWidthHeight(const float4 &box) {
+    return {
+      0.5f * (box.x + box.z),
+      0.5f * (box.y + box.w),
+      box.z - box.x,
+      box.w - box.y};
+}
+
+void BoxEncoder<GPUBackend>::PrepareAnchors(const vector<float> &anchors) {
+  DALI_ENFORCE(
+    (anchors.size() % BoundingBox::kSize) == 0,
+    "Anchors size must be divisible by 4, actual value = " + std::to_string(anchors.size()));
+
+  anchors_count_ = anchors.size() / BoundingBox::kSize;
+  anchors_.Resize({anchors_count_, BoundingBox::kSize});
+  anchors_as_center_wh_.Resize({anchors_count_, BoundingBox::kSize});
+
+  auto anchors_data_cpu = reinterpret_cast<const float4 *>(anchors.data());
+
+  vector<float4> anchors_as_center_wh(anchors_count_);
+  for (unsigned int anchor = 0; anchor < anchors_count_; ++anchor)
+    anchors_as_center_wh[anchor] = ToCenterWidthHeight(anchors_data_cpu[anchor]);
+
+  auto anchors_data = anchors_.mutable_data<float>();
+  auto anchors_as_center_wh_data = anchors_as_center_wh_.mutable_data<float>();
+  MemCopy(anchors_data, anchors.data(), anchors_count_ * BoundingBox::kSize * sizeof(float));
+  MemCopy(
+    anchors_as_center_wh_data,
+    anchors_as_center_wh.data(),
+    anchors_count_ * BoundingBox::kSize * sizeof(float));
+}
+
+__device__ float CalculateIou(const float4 &b1, const float4 &b2) {
+  float l = max(b1.x, b2.x);
+  float t = max(b1.y, b2.y);
+  float r = min(b1.z, b2.z);
+  float b = min(b1.w, b2.w);
+  float first = max(r - l, 0.0f);
+  float second = max(b - t, 0.0f);
+  volatile float intersection = first * second;
+  volatile float area1 = (b1.w - b1.y) * (b1.z - b1.x);
+  volatile float area2 = (b2.w - b2.y) * (b2.z - b2.x);
+
+  return intersection / (area1 + area2 - intersection);
+}
+
+__device__ void FindBestMatch(int N, volatile float *vals, volatile int *idx) {
+  for (unsigned int stride = blockDim.x / 2; stride > 0; stride >>= 1) {
+    if (threadIdx.x < stride) {
+      if (vals[threadIdx.x] <= vals[threadIdx.x + stride]) {
+        if (vals[threadIdx.x] == vals[threadIdx.x + stride]) {
+          idx[threadIdx.x] = max(idx[threadIdx.x], idx[threadIdx.x + stride]);  // NOLINT
+          continue;
+        }
+
+        vals[threadIdx.x] = vals[threadIdx.x + stride];
+        idx[threadIdx.x] = idx[threadIdx.x + stride];
+      }
+    }
+    __syncthreads();
+  }
+}
+
+__device__ void WriteMatchesToOutput(
+  int anchors_count, float criteria, int *labels_out, const int *labels_in,
+  float4 *boxes_out, const float4 *boxes_in,
+  volatile int *best_box_idx, volatile float *best_box_iou) {
```

Comment:
`inline`
Why is bext_box_iou `volatile` here?

Reply:
Same for `bext_box_idx`

Reply:
This encoding is very fragile to floating point arithmetic and one of the requirements was for this to result in exactly the same result as cpu box encoder and reference implementation in python. After a lot of time on this, thoroughly "volatalizing" stuff seems to do the trick, but I agree - it is not the most obvious solution. For know, I would like to move on, since we have some other staff with SSD to fix and then, maybe I can come back to it and  figure out something better. I suspect, that without volatile compiler does some optimizations and results differ (as far as I know, things like this can happen on CPU).

Reply:
perhaps you could add a comment in the code explaining this

Reply:
Done